tests/main/parallel-install-basic/task.yaml

```diff
@@ -0,0 +1,89 @@
+summary: Checks for parallel installation of a local snap files
+
+prepare: |
+    # ensure we have no snap user data directory yet
+    rm -rf /home/test/snap
+    rm -rf /var/snap/test-snapd-tools /var/snap/test-snapd-tools_foo
+
+    snap set system experimental.parallel-instances=true
+
+execute: |
+    #shellcheck source=tests/lib/snaps.sh
+    . "$TESTSLIB"/snaps.sh
+
+    devmode=
+    if [[ "$SPREAD_SYSTEM" = ubuntu-* ]]; then
+        # TODO prallel-install: disable devmode once AppArmor profile updates are in
+        devmode=--devmode
+    fi
+
+    install_local test-snapd-tools
+    install_local_as test-snapd-tools test-snapd-tools_foo $devmode
+
+    su -l -c '! test -d ~/snap/test-snapd-tools' test
+    su -l -c '! test -d ~/snap/test-snapd-tools_foo' test
+
+    su -l -c 'test-snapd-tools_foo.cmd echo foo' test | MATCH foo
+    su -l -c 'test -d ~/snap/test-snapd-tools' test
+    su -l -c 'test -d ~/snap/test-snapd-tools_foo' test
+
+    # instance environment variables are correctly set up
+    su -l -c 'test-snapd-tools_foo.env' test > snap_foo-env.txt
+    MATCH 'SNAP_INSTANCE_NAME=test-snapd-tools_foo'                      < snap_foo-env.txt
+    MATCH 'SNAP_NAME=test-snapd-tools'                                   < snap_foo-env.txt
+    MATCH 'SNAP_INSTANCE_KEY=foo'                                        < snap_foo-env.txt
+    MATCH 'SNAP_COMMON=/var/snap/test-snapd-tools/common'                < snap_foo-env.txt
+    MATCH 'SNAP_DATA=/var/snap/test-snapd-tools/x1'                      < snap_foo-env.txt
```

Comment:
I wonder what bugs we will find when applications are looking at $SNAP_NAME in the environment (which doesn't have _instance here) and expecting it to match @{SNAP_NAME} (which will need to have _instance) in the apparmor profile. Eg, considering the future change to the apparmor profile:
```
@{SNAP_NAME}="test-snapd-tools_foo"
...
profile "snap.test-snapd-tools_foo.env" {
    ...
    /{dev,run}/shm/snap.@{SNAP_NAME}.** mrwlkix,
    ...
```

Eg, an application that added a check for checking environ for $SNAP_NAME might adjust its filename to use `/run/shm/snap.test-snapd-tools/something` when it only has access to `/run/shm/snap.test-snapd-tools_foo/*`. This might affect any apparmor rule that uses SNAP_NAME where the bind mounts this PR performs don't apply (ie, there is no problem with /snap/test-snapd-tools and /var/snap/test-snapd-tools, but there is for all other meditions that reference @{SNAP_NAME}).

Now, if we say that the parallel install feature requires that snaps may need to adjust their snap-specific code (ie, they can no longer naively look at $SNAP_NAME) for a snap to be parallel-installable, then there is no problem with this (non-parallel installs won't break after all, only when someone tries to use the feature might it break). If we say something different, then changes may need to be made to this PR.

Reply:
> Now, if we say that the parallel install feature requires that snaps may need to adjust their snap-specific code (ie, they can no longer naively look at $SNAP_NAME) for a snap to be parallel-installable, then there is no problem with this (non-parallel installs won't break after all, only when someone tries to use the feature might it break).

@niemeyer what do you think about that?

Reply:
We have just discussed this in a larger crowd. We have an agreement that snaps that are expected to be parallel installed need to be aware of the fact. On this ground, using `$SNAP_NAME` naively and fiddling with `/dev/shm/snap.${SNAP_NAME}` is expected to cause trouble with other instances of the same snap stepping over their directories. In that regard, snaps supporting parallel installation should take care to use `${SNAP_INSTANCE_NAME}` when building paths under shared directories (`/dev`, `/run`) themselves. Keep in mind, the usual locations to snap data, snap, user data, `$HOME`, `$XDG_RUNTIME_DIR` are already set up in a way that prevents snaps from stepping over other instances of the same snap.